torchdynamo/guards.py

```diff
@@ -268,12 +268,19 @@ def DICT_KEYS(self, guard):
         self.code.append(f"___check_type_id({ref}, {self.id_ref(type(value))})")
         self.code.append(f"{ref}.keys() == {set(value.keys())!r}")
 
+
     def NN_MODULE_PARAM_NAMES(self, guard):
         ref = self.arg_ref(guard)
         value = self.get(guard.name)
         keys = {k for k, v in value.named_parameters()}
+        values = {v for k, v in value.named_parameters()}
         self.code.append(f"___check_type_id({ref}, {self.id_ref(type(value))})")
-        self.code.append(f"{{k for k, v in {ref}.named_parameters()}} == {keys!r}")
+        self.code.append(f"{{k for k, v in {ref}.named_parameters()}} == {keys!r}")     
+        # TODO after draft diff, before push - make a config value for this?   
+        values_collapsed = f"{values}".replace(' ', '').replace('\n', '')
+        self.code.append(f"{{v for k, v in {ref}.named_parameters()}} == {values_collapsed!r}")
```

Comment:
Guarding on exact tensor values like this will:
1) Be very expensive, since it is constructing a Tensor inline.  (Which could result in a guard string of possibly megabytes in size.)
2) Will cause this guard to fail every iteration in training, because the parameters are being learned and change.
3) I think this guard will just always fail:
```
>>> value=torch.nn.Linear(2,2)
>>> values = {v for k, v in value.named_parameters()}
>>> values_collapsed = f"{values}".replace(' ', '').replace('\n', '')
>>> ref = "value"
>>> code = f"{{v for k, v in {ref}.named_parameters()}} == {values_collapsed!r}"
>>> eval(code)
False
>>> code
"{v for k, v in value.named_parameters()} == '{Parametercontaining:tensor([[0.6959,0.2447],[-0.5361,0.3697]],requires_grad=True),Parametercontaining:tensor([0.5704,0.1800],requires_grad=True)}'"
>>> {v for k, v in value.named_parameters()} == '{Parametercontaining:tensor([[0.6959,0.2447],[-0.5361,0.3697]],requires_grad=True),Parametercontaining:tensor([0.5704,0.1800],requires_grad=True)}'
False
>>> {v for k, v in value.named_parameters()}
{Parameter containing:
tensor([[ 0.6959,  0.2447],
        [-0.5361,  0.3697]], requires_grad=True), Parameter containing:
tensor([0.5704, 0.1800], requires_grad=True)}
>>> 


```
Note that `eval(code)` returned false even for an identical `value`

For a tensor on an NN_MODULE perhaps we want to use `ID_MATCH` instead.  That will check to see if the user is swapping out parameters dynamically, which will confuse things as NNModuleVariable assumes that doesn't happen.  If that is the case, we may want to try to switch the class in question to [UnspecializedNNModuleVariable](https://github.com/pytorch/torchdynamo/blob/063bc13ee2e9f41e25e4b4ac96a62adcb944697b/torchdynamo/variables/nn_module.py#L327).

Reply:
Great suggestions, let me try.

And yes, this is prohibitively expensive.